pandas/_libs/src/vendored/ujson/python/objToJSON.c

```diff
@@ -467,12 +474,19 @@ static int NpyArr_iterNextItem(JSOBJ obj, JSONTypeContext *tc) {
 
   NpyArr_freeItemValue(obj, tc);
 
-  if (PyArray_ISDATETIME(npyarr->array)) {
+  if (!PyArray_Check(npyarr->array)) {
+    PyErr_SetString(PyExc_TypeError,
+                    "NpyArr_iterNextItem received a non-array object");
+    return 0;
+  }
+  PyArrayObject *arrayobj = (PyArrayObject *)npyarr->array;
```

Comment:
This is the only instance in our JSON code where we could not use a `const` qualifier for the `PyArrayObject` type because `PyArray_DESCR` accepts a non-const type. I think this is something that will eventually be addressed in https://github.com/numpy/numpy/issues/24855#issue-1926037228 but @seberg probably knows more

Reply:
Yeah, I noticed that problem as well today.  I think we should change it, might annoy some C++ folks, but should be easy enough to work around.